File: lib/client/player.ts
import { cache } from '@communityox/ox_lib/client';
import type { OxPlayer } from 'client/player';
import type { Dict } from 'types';

class PlayerInterface {
  public userId: number;
  public charId?: number;
  public stateId?: string;
  [key: string]: any;

  constructor() {
    try {
      const { userId, charId, stateId } = exports.ox_core.GetPlayer();
      this.userId = userId;
      this.charId = charId;
      this.stateId = stateId;
    } catch (e) {}

    this.state = LocalPlayer.state;

    this.constructor.prototype.toString = () => {
      return JSON.stringify(this, null, 2);
    };

    const getMethods = async () => {
      Object.keys(exports.ox_core.GetPlayerCalls()).forEach((method: string) => {
        if (!this.constructor.prototype[method])
          this.constructor.prototype[method] = (...args: any[]) => exports.ox_core.CallPlayer(method, ...args);
      });
    };

    // Prevent errors if resource starts before ox_core (generally during development)
    getMethods().catch(() => setImmediate(getMethods));
  }

  /**
   * Registers an event handler which will be triggered when the specified player data is updated.
   */
  on(key: string, callback: (data: unknown) => void) {
    this.get(key);

    on(`ox:player:${key}`, (data: unknown) => {
      if (GetInvokingResource() == 'ox_core' && (source as any) === '') callback(data);
    });
  }

  /**
   * Returns player data for the specified key. The data is cached and kept updated for future calls.
   */
  get(key: string) {
    if (!this.charId) return;

    if (!(key in this)) {
      this[key] = exports.ox_core.CallPlayer('get', key) ?? null;
      this.on(key, (data: unknown) => (this[key] = data));
    }

    return this[key];
  }

  getCoords() {
    return GetEntityCoords(cache.ped);
  }
}

export type OxPlayer = typeof OxPlayer & PlayerInterface;

const player = new PlayerInterface() as OxPlayer;

export function GetPlayer() {
  return player;
}

on('ox:playerLoaded', (data: Dict<any>) => {
  if (player.charId) return;

  for (const key in data) player[key] = data[key];
});

on('ox:playerLogout', () => {
  for (const key in player) delete player[key];
});

File: lib/client/player.lua
---@diagnostic disable: redundant-parameter
---@class OxPlayerClient : OxClass
local OxPlayer = lib.class('OxPlayer')

-- Support for `player.method` rather than self (:) syntax
function OxPlayer:__index(index)
    local value = OxPlayer[index] --[[@as any]]

    if type(value) == 'function' then
        self[index] = value == OxPlayer.__call and function(...)
            return value(self, index, ...)
        end or function(...)
            return value(self, ...)
        end

        return self[index]
    end

    return value
end

function OxPlayer:constructor()
    pcall(function()
        local data = exports.ox_core.GetPlayer()

        self.userId = data.userId
        self.charId = data.charId
        self.stateId = data.stateId
    end)

    self.state = LocalPlayer.state
end

function OxPlayer:__call(...)
    return exports.ox_core:CallPlayer(...)
end

function OxPlayer:__tostring()
    return string.format('{\n  "userId": %s\n  "charId": %s\n  "stateId": %s\n}',
        self.userId, self.charId, self.stateId)
end

local getters = {}

function OxPlayer:on(key, callback)
    self.get(key)

    AddEventHandler(('ox:player:%s'):format(key), function(data)
        if GetInvokingResource() == 'ox_core' and source == '' then
            callback(data)
        end
    end)
end

function OxPlayer:get(key)
    if not self.charId then return end

    if not getters[key] then
        getters[key] = true

        self.on(key, function(data) self[key] = data end)
        self[key] = OxPlayer:__call('get', key);
    end

    return self[key]
end

function OxPlayer:getCoords()
    return GetEntityCoords(cache.ped);
end

function OxPlayer:getGroup(filter)
    local result = OxPlayer:__call('getGroup', filter)

    if type(result) == 'table' then
        return table.unpack(result)
    end

    return result
end

function OxPlayer:getGroupByType(type)
    local result = OxPlayer:__call('getGroupByType', type)

    if result then
        return table.unpack(result)
    end
end

---@class OxClient
local Ox = Ox

local player = OxPlayer:new()

function Ox.GetPlayer()
    return player
end

local function getMethods()
    for method in pairs(exports.ox_core:GetPlayerCalls()) do
        if not rawget(OxPlayer, method) then OxPlayer[method] = OxPlayer.__call end
    end
end

-- Prevent errors if resource starts before ox_core (generally during development)
if not pcall(getMethods) then CreateThread(getMethods) end

AddEventHandler('ox:playerLoaded', function(data)
    if player.charId then return end

    for k, v in pairs(data) do player[k] = v end
end)

AddEventHandler('ox:playerLogout', function()
    table.wipe(player)
end)
